8336906: C2: assert(bb->is_reachable()) failed: getting result from unreachable basicblock

mhaessig opened the pull request:
# Issue Summary

This PR addresses an `assert(bb->is_reachable())` that is triggered in the code for `-XX:+VerifyStack` after a deoptimization with reason `null_assert_or_unreached0` at a `getstatic` bytecode. Following the `getstatic` is an `areturn` and then an unreachable bytecode. When the code for `VerifyStack` tries to compute an oop map for the basic block of the unreachable bytecode, the assert triggers:
```
getstatic Field A.val:"LB"; // if class B is not loaded, C2 deopts with reason "null_assert_or_unreached0"
areturn;
// The following is unreachable
iconst_0;
```

This is a similar problem to [JDK-8271055](https://bugs.openjdk.org/browse/JDK-8271055) (#7331), but this particular deopt with reason `null_assert_or_unreached0` at `getstatic` of a field containing an object reference [deopts at the next bytecode](https://github.com/openjdk/jdk/blob/ad07426fab3396caefd7c08d924e085c1f6f61ba/src/hotspot/share/opto/parse3.cpp#L176-L199). The aforementioned issue introduced a check to skip stack verification of the next bytecode in the code if the execution after the deopted bytecode does not continue at the next bytecode in the code, i.e. falls through to the next bytecode.  Unfortunately, this check did not include `areturn` as a bytecode that does not fall-through:
https://github.com/openjdk/jdk/blob/ad07426fab3396caefd7c08d924e085c1f6f61ba/src/hotspot/share/runtime/deoptimization.cpp#L845-L856

# Change Summary

To fix the immediate issue described above, this PR adds `areturn` to the list of bytecodes that does not fall through. However, all return bytecodes exhibit the same behavior and might be susceptible to a similar issue. Even though I was not able to reproduce the same crash with `{d,f,i,l}return` because I could not get those or the preceding bytecode to deopt, I also added them to the `falls_through()` function. For the remaining bytecodes in `falls_through()` with the exception of `athrow` I wrote a regression test.

# Testing

- [x] [Github Actions](https://github.com/mhaessig/jdk/actions/runs/14595928439)
- [x] tier1 through tier3 on Oracle supported platforms and OSs plus Oracle internal testing

# Acknowledgements
Special thanks to @eme64 for his hard work on reducing a reproducer that works on all platforms.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336906](https://bugs.openjdk.org/browse/JDK-8336906): C2: assert(bb-&gt;is_reachable()) failed: getting result from unreachable basicblock (**Bug** - P4)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Contributors
 * Emanuel Peter `<epeter@openjdk.org>`
 * Dean Long `<dlong@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25118/head:pull/25118` \
`$ git checkout pull/25118`

Update a local copy of the PR: \
`$ git checkout pull/25118` \
`$ git pull https://git.openjdk.org/jdk.git pull/25118/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25118`

View PR using the GUI difftool: \
`$ git pr show -t 25118`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25118.diff">https://git.openjdk.org/jdk/pull/25118.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25118#issuecomment-2863082490)
</details>

mhaessig commented:
/contributor add @eme64

mhaessig commented:
/cc hotspot-compiler

eme64 commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -842,13 +842,21 @@ void Deoptimization::unwind_callee_save_values(frame* f, vframeArray* vframe_arr
+    case Bytecodes::_ireturn:
+    case Bytecodes::_lreturn:
+    case Bytecodes::_ret:
+    case Bytecodes::_return:
```
Did you go through all bytecodes we support here? I quickly scanned the wiki page, and found the (depricated but still present) `jsr` opcodes that also are essencially a `goto`, so do not fall-through, I think. At least it seems to me we could also have unreachable code after a `jsr`, right?

Can you go over all bytecodes, and make sure we are not missing any? Because this is already the second bug of this kind, would be good to fix it once and for all now ;)

eme64 reviewed (changes requested):
@mhaessig Thanks for looking into this!

A few comments:

1)
Did you go through all bytecodes we support here? I quickly scanned the wiki page, and found the (depricated but still present) `jsr` opcodes that also are essencially a goto, so do not fall-through, I think. At least it seems to me we could also have unreachable code after a `jsr`, right? And then there is a `ret` bytecode that does the symmetrical thing. Not sure if we even handle these in such a way that the bug could be reproduced, but worth a check!

Can you go over all bytecodes, and make sure we are not missing any? Because this is already the second bug of this kind, would be good to fix it once and for all now ;)

2)
> Even though I was not able to reproduce the same crash with {d,f,i,l}return because I could not get those or the preceding bytecode to deopt, I also added them to the falls_through() function.

Hmm ok, I see. Might be worth investing just a little more time to see if we cannot get that done. Or else argue why it CANNOT be done. But then we might as well put an assert inside `falls_through` for those cases, to check that we actually never deopt like that at such an opcode, and revisit that assumption if we ever hit the assert. What do you think?

3)
For the test: It's a bit of a shame to have lots of separate files. Especially because the directory name `test/hotspot/jtreg/compiler/interpreter/verifyStack/` may at some point have more tests, and then it gets a little confusing. I wonder if `A` and `B` could be nested classes in the java file? And the java and jasm file could be named very similarly, so that it is directly clear that they belong together when browsing the test files. An alternative: create a subdirectory that has a very unique name, so that we could separate things that way.

eme64 commented:
> Even though I was not able to reproduce the same crash with {d,f,i,l}return because I could not get those or the preceding bytecode to deopt, I also added them to the falls_through() function.

Basically, there are 2 cases:
- opcodes that deopt and retry: these were already there, as far as I know, and @dean-long added them in his previous patch. So here we could only take opcodes that: deopt, retry, and do not have fall-through.
- opcodes that deopt but do not retry, but skip forward to the next op, that we then have to check for fall-through. For the deopting opcode, there are 2 categories:
  - Those that put something on the stack, like `getstatic` that puts whatever it got on the stack. This constrains what opcode comes after. If it returns an object/null, you can only do `return` (ignore stack value) or `areturn` (return that stack value). But you cannot do `ireturn` because the value on the stack is an object, not int.
  - Those that put nothing on the stack. Here we would not be constrained, and could push whatever we need on the stack before that opcode. E.g. we could push an int before that opcode, and then do `ireturn`. But I'm not sure if there are any such opcodes that deopt but push nothing on the stack. Worth checking though!

Hope this case distinction helps a little, I'm not sure it is particularly clear or accurate, but these are the things I would look into if I were working on this bug :)

eme64 commented:
Quickly scanning, I see these that also may or may not have a fall-through: `lookupswitch`, `tableswitch`

dean-long commented:
Making falls_through() handle all cases, including lookupswitch and tableswitch seems like the right fix.  When I added it originally, I was not aware that C2 could set the bci to the next instruction instead of the current instruction.  I think this means almost any instruction could be encountered at the "next next" bci.

mhaessig commented:
> For the test: It's a bit of a shame to have lots of separate files.

I got myself confused with class visibilities. I moved the `A` and `B` into the java test file and moved both of them one directory lower into `compiler/interpreter`. That makes it a lot cleaner.

> Or else argue why it CANNOT be done.

I looked into it some more and there are two places where we deopt and move the `bci` over to the next bytecode: when an object of an unloaded class is returned by `getstatic` (see code in the PR description) and calls (see below) and the object reference is `null`.

https://github.com/openjdk/jdk/blob/411a63ea1b0c6e8bfea219427bf1c317c5dadabf/src/hotspot/share/opto/doCall.cpp#L770-L785

Since `{d,f,i,l}return`, `{table,lookup}switch`, and `ret` require an integer on the stack but only bytecodes that push references on the stack can deopt to the next `bci`, we cannot trigger this error for those bytecodes. Now, that begs the question, whether these bytecodes should then be in `falls_through()`. I argue that they should be, since that would be the correct behavior if we deopted at such a bytecode.

> Did you go through all bytecodes we support here?

I did, but I missed `lookupswitch` and `tableswitch`. `jsr` just pushes the address of the next bytecode onto the stack. `ret` can jump to such an address, but I already added that to `falls_through()`.

dean-long commented:
BTW, I think determining whether the instruction after a `jsr` is reachable is not so easy.  Normally if the subroutine does a `ret` then we return to the instruction after the jsr, but I think the subroutine can also do things like throw an exception or loop forever, so it's probably better to treat `jsr` the same as `goto` and add it to `falls_through`.  

But now that I think about it, I'm wondering if this special logic is still needed by VerifyStack.  It might not be needed at all, or it might only be needed for specific situations, like an invoke instruction or an instruction with the reexecute flag set.

dean-long commented:
```
957            // The interpreter oop map generator reports results before
958            // the current bytecode has executed except in the case of
959            // calls. It seems to be hard to tell whether the compiler
960            // has emitted debug information matching the "state before"
961            // a given bytecode or the state after, so we try both
```
The comment justifying this logic claims the compiler can emit debug information with the "after" state instead of the "before" state.  If this is not true, then we can remove this VerifyStack logic.  If it is true, it would be good to understand exactly under what circumstances it can happen.

dean-long commented:
I forgot that uncommon traps always use reexecute semantics, even if they don't explicitly set the reexecute flag on the debug info.  That's why they can update the JVM state and advance to the next bci.  It seems like it would be safe to skip the falls_through/try_next_mask logic for the top frame if we are reexecuting at the current bci, but that is a riskier change.

BTW, because some uncommon traps advance to the next bci, code like Deoptimization::gather_statistics() will not be gathering statistics for the bytecode that causes the trap, making the statistics less meaningful.  It seems better to allow uncommon traps to use the current bci instead and let the deoptimization code advance it if the reexecute flag is not set.

mhaessig commented:
> so it's probably better to treat jsr the same as goto and add it to falls_through.

I added `jsr` to `falls_through()`.

> The comment justifying this logic claims the compiler can emit debug information with the "after" state instead of the "before" state. If this is not true, then we can remove this VerifyStack logic. If it is true, it would be good to understand exactly under what circumstances it can happen.

Isn't this just what is happening in this particular bug? But it is not the debug info that is emitted before or after and only the `bci` that is rarely advanced. Granted, I have not been able to look into that in depth, because I am unfamiliar with those debuginfo emitting codepaths, but a cursory look seemed to suggest that it is always emitted at the end of `PhaseOutput::Process_OopMap_Node` which would be the same for all bytecodes.

> It seems like it would be safe to skip the falls_through/try_next_mask logic for the top frame if we are reexecuting at the current bci, but that is a riskier change.

I agree, but I don's see a good way to find out if the `bci` has been moved. A hacky solution might be to check if the deopt reason is `assert_null_or_unreachable0` since the `bci` is only moved in some deopts with that reason.

dean-long commented:
What makes the debug info "before" or "after" is the state of the stack relative to the bci.  So if we are doing a getfield, and have already popped the object off the stack and pushed the result, then the state would be "after".  However, we then advance the bci, making the state "before", because we are now on a different bytecode.  Since uncommon traps have reexecute semantics, the interpreter will continue executing at that bci without advancing it first, so we need a "before" state in that case.  That's why trying to compute an "after" state doesn't make sense to me for uncommon traps.  They should all use a "before" state. 
In other deoptimization cases without reexecute semantics, we may cause the interpreter to advance to the bci that follows the one given in the debug info.  In this situation, the debug info would need to be the "after" state, and bci in the debug info would need to be one that falls through.
I plan to take a closer look at this VerifyStack code and possibly improve or remove it, but not as part of this PR.

eme64 commented on src/hotspot/share/runtime/deoptimization.cpp:
```diff
@@ -842,13 +842,24 @@ void Deoptimization::unwind_callee_save_values(frame* f, vframeArray* vframe_arr
 }
 
 #ifndef PRODUCT
+// Return true if the execution after the provided bytecode continues at the
```
```suggestion
// Return true if the execution after the provided bytecode can continue at the
```
Nit: Because a `cmp_if` may or may not continue with the next bci.

mhaessig replied:
I looked into that and both basic blocks following the `if_cmp` will always be reachable as far as `bb->is_reachable()` is concerned. The bytecode verification code of the `if_cmp` to check that the stack depth on both the then and else branch are the same requires both basic blocks to be reachable.

dean-long replied:
Yes, the function name and comment aren't as precise as they could be.  "parsing" might have been a better word than "execution", for example.  But I think the function name is OK as it is.  We use variants of this naming elsewhere for exactly the same purpose:
https://github.com/openjdk/jdk/blob/e7ce661adb01fba4bb690d51cc2858c822008654/src/hotspot/share/oops/generateOopMap.cpp#L1159

eme64 reviewed (approved):
Looks reasonable to me :)

TobiHartmann commented:
/sponsor
